parquet/src/main/java/org/apache/iceberg/parquet/BasePageIterator.java

```diff
@@ -0,0 +1,202 @@
+ */
+
+package org.apache.iceberg.parquet;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import org.apache.parquet.bytes.ByteBufferInputStream;
+import org.apache.parquet.bytes.BytesInput;
+import org.apache.parquet.bytes.BytesUtils;
+import org.apache.parquet.column.ColumnDescriptor;
+import org.apache.parquet.column.Dictionary;
+import org.apache.parquet.column.Encoding;
+import org.apache.parquet.column.ValuesType;
+import org.apache.parquet.column.page.DataPage;
+import org.apache.parquet.column.page.DataPageV1;
+import org.apache.parquet.column.page.DataPageV2;
+import org.apache.parquet.column.values.ValuesReader;
+import org.apache.parquet.column.values.rle.RunLengthBitPackingHybridDecoder;
+import org.apache.parquet.io.ParquetDecodingException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("checkstyle:VisibilityModifier")
+public abstract class BasePageIterator {
+  private static final Logger LOG = LoggerFactory.getLogger(BasePageIterator.class);
+
+  protected final ColumnDescriptor desc;
+  protected final String writerVersion;
+
+  // iterator state
+  protected boolean hasNext = false;
+  protected int triplesRead = 0;
+  protected int currentDL = 0;
+  protected int currentRL = 0;
+
+  // page bookkeeping
+  protected Dictionary dictionary = null;
+  protected DataPage page = null;
+  protected int triplesCount = 0;
+  protected Encoding valueEncoding = null;
+  protected IntIterator definitionLevels = null;
+  protected IntIterator repetitionLevels = null;
+  protected ValuesReader vectorizedDefinitionLevelReader = null;
+  protected ValuesReader values = null;
+
+  protected BasePageIterator(ColumnDescriptor descriptor, String writerVersion) {
+    this.desc = descriptor;
+    this.writerVersion = writerVersion;
+  }
+
+  protected abstract void reset();
+
+  protected abstract boolean supportsVectorizedReads();
+
+  protected abstract IntIterator newNonVectorizedDefinitionLevelReader(ValuesReader dlReader);
+
+  protected abstract ValuesReader newVectorizedDefinitionLevelReader(ColumnDescriptor descriptor);
+
+  protected abstract void initDataReader(Encoding dataEncoding, ByteBufferInputStream in, int valueCount);
+
+  public void setPage(DataPage page) {
+    Preconditions.checkNotNull(page, "Cannot read from null page");
+    this.page = page;
+    this.page.accept(new DataPage.Visitor<ValuesReader>() {
+      @Override
+      public ValuesReader visit(DataPageV1 dataPageV1) {
+        initFromPage(dataPageV1);
+        return null;
+      }
+
+      @Override
+      public ValuesReader visit(DataPageV2 dataPageV2) {
+        initFromPage(dataPageV2);
+        return null;
+      }
+    });
+    this.triplesRead = 0;
+    this.hasNext = triplesRead < triplesCount;
+  }
+
+  protected void initFromPage(DataPageV1 initPage) {
+    this.triplesCount = initPage.getValueCount();
+    ValuesReader dlReader = null;
+    if (supportsVectorizedReads()) {
```

Comment:
Couldn't this create a vectorized definition level reader either way? That still exposes `nextInteger()`, which is all the non-vectorized version needs to call for an `IntIterator`. Then you could always use the same DL reader and the subclasses can choose how to wrap it (non-vectorized) or call its methods (vectorized).

Reply:
Another way to clean up this and its subclasses is to use a method like the one for value reader initialization, `initDataReader`. Couldn't this call `initDefinitionLevelReader` with `triplesCount`, `in`, and `desc`? Then the subclasses would be responsible for building and initializing their own DL reader.